src/Symfony/Component/DependencyInjection/Compiler/ResolveBindingsPass.php

```diff
@@ -0,0 +1,123 @@
+
+        $parameterBag = $this->container->getParameterBag();
+
+        if ($class = $value->getClass()) {
+            $class = $parameterBag->resolveValue($class);
+        }
+
+        $bindings = $value->getBindings();
+
+        $calls = $value->getMethodCalls();
+        // constructor
+        $calls[] = array(null, $value->getArguments());
+
+        foreach ($calls as $i => $call) {
+            list($method, $arguments) = $call;
+            $method = $parameterBag->resolveValue($method);
+            $parameters = null;
+            $resolvedArguments = array();
+
+            if (null === $reflectionMethod = $this->getReflectionMethod($class, $method)) {
+                continue;
+            }
+
+            foreach ($reflectionMethod->getParameters() as $key => $parameter) {
+                if (isset($arguments[$key])) {
+                    continue;
+                }
+
+                $typeHint = ProxyHelper::getTypeHint($reflectionMethod, $parameter, true);
+
+                if (!isset($bindings[$typeHint])) {
+                    continue;
+                }
+
+                $arguments[$key] = $bindings[$typeHint];
+            }
+
+            if ($arguments !== $call[1]) {
+                ksort($arguments);
+                $calls[$i][1] = $arguments;
+            }
+        }
+
+        list(, $arguments) = array_pop($calls);
+
+        if ($arguments !== $value->getArguments()) {
+            $value->setArguments($arguments);
+        }
+        if ($calls !== $value->getMethodCalls()) {
+            $value->setMethodCalls($calls);
+        }
+
+        return parent::processValue($value, $isRoot);
+    }
+
+    /**
+     * @param string|null $class
+     * @param string|null $method null if is the constructor
+     *
+     * @throws InvalidArgumentException
+     *
+     * @return \ReflectionMethod|null
+     */
+    private function getReflectionMethod($class, $method)
+    {
+        if (!$class) {
+            throw new InvalidArgumentException(sprintf('Unable to resolve service "%s": the class is not set.', $this->currentId));
+        }
+
+        if (!$r = $this->container->getReflectionClass($class)) {
+            throw new InvalidArgumentException(sprintf('Unable to resolve service "%s": class "%s" does not exist.', $this->currentId, $class));
+        }
+
+        if (null === $method) {
+            $method = $r->getConstructor();
+            if (null === $method) {
+                return;
+            }
+        } else {
+            if (!$r->hasMethod($method)) {
+                throw new InvalidArgumentException(sprintf('Unable to resolve service "%s": method "%s::%s" does not exist.', $this->currentId, $class, $method));
+            }
+
+            $method = $r->getMethod($method);
+        }
+
+        if (!$method->isPublic()) {
+            throw new InvalidArgumentException(sprintf('Unable to resolve service "%s": method "%s::%s" must be public.', $this->currentId, $class, $method->name));
```

Comment:
I'd remove this check, other passes do it (eg AutowirePass) and it's not the primary responsibility of this pass.